### DISCUSSION TITLE
How to add PodName / Template as labels for a metric

### DISCUSSION THREAD

**Discussion by reenimathew** [argoproj/argo-workflows#8184] [Q&A]
Hi Team,

We are trying to introduce an ability to find the ResourceUsage / Duration of any workflow.  We like to leverage metrics endpoint scrape to fetch all necessary information (Finding this as the only solution to extract information).

And these are the labels we are looking for.
1) Namespace
2) WorkflowName
3) PodName
4) Template

**Expectation:**
argo_workflows_exec_duration_cpu{ name="hello-world-wf-global-arg-sv7lv",namespace="argo", podName="hello-world-wf-global-arg-48xff-2243349284", template="workflow-template-submittable/whalesay-template"}

**Usecases:**
1) Get ResourceUsage at top level workflow
    Labels required: Namespace, Template, WorkflowName

2) Get ResourceUsage at individual workflow inside main workflow
    Labels required: Namespace, Template, WorkflowName, PodName

**Present prometheus metric**
`argo_workflows_exec_duration_cpu{instance="X.X.X.X:9090", job="argo-prometheus", name="hello-world-wf-global-arg-sv7lv", namespace="argo", node="{{ARGO_NODE_ID}}"}`

![image](https://user-images.githubusercontent.com/6531836/158938620-dba27d24-d145-48ba-8515-b76527e7822b.png)

**Complete workflow template available here:**
https://github.com/reenimathew/argo-basics/blob/main/metric/wftemplatemain.yaml

**Challenges faced**
1) Not able to find a way to pick ResourceDuration at complete workflow level (Alternate solution - Aggregate individual workflows given, podName is available at individual workflow level)
2) Not able to find a way to provide podName as the Label for every metric
3) Not able to find a way to provide template as the Label for every metric

Getting exception when trying to access NodeId annotation which is present by default in every pod description. Not sure whether its a right way to do this.

[argo]$ argo submit --namespace argo wftemplatemain.yaml --serviceaccount argo
FATA[2022-03-18T07:45:10.392Z] Failed to submit workflow: templates.whalesay.steps[1].hello1 templates.whalesay1: failed to resolve {{workflow.annotations.workflows.argoproj.io/node-id}} 

Could you please let us know, how to get these information.

**Additional Query:**
Is it possible to access ResourceDuration in any lifecycle hook where we can persist it some 3rd party file system like S3 .
Tried from onExit and it was not available. And it makes sense as onExit can also add to complete ResourceDuration.

Thanks
Reeni

**Comment by alexec** [comment]
Should this be an enhancement request?

**Reply by reenimathew** [reply]
If this capability is not available as of now, we can consider this as an enhancement request @alexec 
Is it not possible to get NodeId in a metrics as key as of now?
